docs/mocksupport.rst

```diff
@@ -867,13 +738,338 @@ This adds the instances to the repository display of the previous example:
+
+.. code-block:: python
+
+    def __init__(self, cimrepository):
+        """
+        Init of test provider
+        """
+        super(UserMethodTestProvider, self).__init__(cimrepository)
+
+
+2. Definition of the CIM class the provider supports as a class variable.
+
+.. code-block:: python
+
+    provider_classnames = 'CIM_Foo'
+
+3. Definition and implementation of the WBEM operations supported by the
+   provider from the set of request operations defined in the table above for a
+   specific provider type. This must be a subset of the methods defined for the
+   provider type. Methods that do not exist in the user-defined provider
+   default to the default method. Each of these methods may:
+
+   * provide parameter or CIM repository validation in addition to the normal
+     request validation,
+
+   * modify parameters of the request.
+
+   * generate an CIMError exception for the operation.
+
+   * make modifications to the CIM repository.
+
+   The request may be completed by the user provider method calling the
+   superclass corresponding method to complete its final validation and
+   modify the repository or by the provider directly modifying the repository
+   based on the method and parameters.
+
+    The user-defined providers have access to:
+       * methods defined in their superclass
+       * methods defined in :class:~pywbem_mock.BaseProvider`
+       * methods to access the CIM repository using the methods defined in
+         :class:`~pywbem_mock.InMemoryRepository`
+
+   TODO define the aprori validation for each operation. Good place for table
+
+.. code-block:: python
+
+    # define CreateInstance and call the superclass method
+    def CreateInstance(self, namespace, NewInstance):
+        """Test Create instance just calls super class method"""
+
+        # Simply calls the default responder method for CreateInstance
+        return super(UserInstanceTestProvider, self).CreateInstance(
+            namespace, NewInstance)
+
+    TODO show example where it specifically modifies the repository.
+
+3. Optional method (``post_register-setup()``) that
+   :meth:`~pywbem_mock.FakedWBEMConnection.register_provider` will
+   call as after the provider registration is successful.  This allows the
+   provider to do any special setup it desires uses its own methods. This
+   method includes the current connection as a parameter so that client
+   methods can be executed directly.
+
+.. code-block:: python
+
+    def post_register_setup(self, conn):
+        ....
+
+
+.. _`Creating instance user-defined providers`:
+
+Creating instance user-defined providers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Instance user-defined methods can override the default implementations for the
+methods defined in the above table. These methods are defined in a user-defined
+class that subclasses :class:`pywbem_mock.InstanceWriteProvider`. The details
+of defining this provider are in section :ref:`Instance Write Provider`
```

Comment:
I think this is too complicated. We describe nearly everything here, and then again in the `_instancewriteprovider` module. I suggest to have the description for each provider type only in the corresponding module. Any example code can be shown in the module docstring.

This also applies ot the method provider.

Reply:
Removed the whole thing and referenced the method definition.  This was already done for the Method provider.